actionview/lib/action_view/test_case.rb

```diff
@@ -42,6 +42,160 @@ def initialize
+        end
+
+        # Register a callable to decode rendered content for a given MIME type
+        #
+        # Each registered decoder will also define a <tt>#rendered_$MIME</tt> helper
+        # method, where <tt>$MIME</tt> corresponds to the value of the
+        # <tt>mime</tt> argument
+        #
+        # === Arguments
+        #
+        # <tt>mime</tt> - Symbol the MIME Type name for the rendered content
+        # <tt>callable</tt> - Callable to decode the String. Accepts the String
+        #                     value as its only argument
+        # <tt>block</tt> - Block serves as the decoder when the
+        #                  <tt>callable</tt> is omitted
+        #
+        # By default, ActionView::TestCase defines a decoder for:
+        #
+        #   * :html - returns an instance of Nokogiri::XML::Node
+        #   * :json - returns an instance of ActiveSupport::HashWithIndifferentAccess
+        #
+        # Each pre-registered decoder also defines a corresponding helper:
+        #
+        #   * :html - defines `rendered_html`
+        #   * :json - defines `rendered_json`
+        #
+        # === Examples
+        #
+        #   test "renders HTML" do
+        #     article = Article.create!(title: "Hello, world")
+        #
+        #     render partial: "articles/article", locals: { article: article }
+        #
+        #     assert_pattern { rendered_html.at("main h1") => { content: "Hello, world" } }
+        #   end
+        #
+        #   test "renders JSON" do
+        #     article = Article.create!(title: "Hello, world")
+        #
+        #     render formats: :json, partial: "articles/article", locals: { article: article }
+        #
+        #     assert_pattern { rendered_json => { title: "Hello, world" } }
+        #   end
+        #
+        # To parse the rendered content into RSS, register a call to <tt>RSS::Parser.parse</tt>:
+        #
+        #   register_decoder :rss, -> rendered { RSS::Parser.parse(rendered) }
+        #
+        #   test "renders RSS" do
+        #     article = Article.create!(title: "Hello, world")
+        #
+        #     render formats: :rss, partial: article
+        #
+        #     assert_equal "Hello, world", rendered_rss.items.last.title
+        #   end
+        #
+        # To parse the rendered content into a Capybara::Simple::Node,
+        # re-register an <tt>:html</tt> decoder with a call to
+        # <tt>Capybara.string</tt>:
+        #
+        #   register_decoder :html, -> rendered { Capybara.string(rendered) }
+        #
+        #   test "renders HTML" do
+        #     article = Article.create!(title: "Hello, world")
+        #
+        #     render partial: article
+        #
+        #     rendered_html.assert_css "h1", text: "Hello, world"
+        #   end
+        def register_decoder(mime, callable = nil, &block)
+          decoder = decoders[mime] = callable || block || :itself.to_proc
+
+          redefine_method "rendered_#{mime}" do
```

Comment:
As an alternative to defining a new method, I also considered redefining `rendered` to be an object that wraps the String.

It'd delegate all missing calls to the underling String, but would also define methods like `#html` and `#json` to do the parsing. Tests would call `rendered.html`, `rendered.json`, etc.

Existing tests that assumed it was a String (like `assert_includes`, `assert_match`, etc) might continue to work.

The biggest backwards compatibility risk would be for tests that did class checking or passed the String to other methods.

Would that pattern be preferable to defining methods with the format suffix in their name?

Reply:
I think this is the biggest risk of this PR, if we are defining methods in users code it can easily create confusing scenarios. Like if someone already defines a `rendered_whatever` then this code fails right?

This is just me thinking aloud though, if we introduce this feature that _might_ break some applications, if we can make it opt-in that is probably better than introducing possibly confusing bugs where we redefine methods? Unless the latter can also be optional, and safe-guard against those scenarios.

Reply:
@zzak thats a great point. For backwards compatibility sake, I wonder if we could introduce a `config.action_view` value that's `true` for 7.1 apps and `false` for older apps.

With that guard to force opt-in, we could be more ambitious with the module's interface, and could override `rendered` to return an instance of a class that responds to the format methods (`rendered.html`, `rendered.json`, etc.) while delegating missing methods to the `String` content.

That way, older apps, apps that haven't opted-in, or apps that have opted-out could continue to rely on `rendered` returning a `String`.

Reply:
I always try to avoid prescribing compatibility-like work before an idea is accepted, but I think you're on the right track.

Reply:
@zzak I've taken your feedback into account, replaced the `rendered_html` with `rendered.html`. I've also introduced the `config.action_view.parse_rendered_content_in_tests` configuration. It defaults to `false` prior to 7.1, and `true` in 7.1.